src/vocabulary/locator.rs

```diff
@@ -0,0 +1,238 @@
+use hf_hub::{api::sync::ApiBuilder, Repo, RepoType};
+use serde::{Deserialize, Serialize};
+use tokenizers::{FromPretrainedParameters, Tokenizer};
+
+use crate::primitives::*;
+
+/// List of common eos token locations appearing on hugging face hub, ordered by priority.
+const COMMON_LOCATIONS: &[EosTokenLocation] = &[
+    // Most projects have `generation_config.json` that looks like:
+    // {
+    //   ...
+    //   "eos_token_id": 50256,
+    //   ...
+    // }
+    // So it's the first place we look for the eos token id.
+    //
+    // For example:
+    // - https://huggingface.co/openai-community/gpt2/blob/main/generation_config.json
+    EosTokenLocation {
+        file: "generation_config.json",
+        location: EosTokenField::Id,
+    },
+    // The ones that don't have `generation_config.json` usually have `tokenizer_config.json`:
+    // {
+    //   ...
+    //   "eos_token": "<|endoftext|>",
+    //   ...
+    // }
+    // Once we have the eos token content, we can get its id from the tokenizer.
+    //
+    // For example:
+    // - https://huggingface.co/microsoft/phi-2/blob/main/tokenizer_config.json
+    EosTokenLocation {
+        file: "tokenizer_config.json",
+        location: EosTokenField::Value,
+    },
+    // Sometimes `tokenizer_config.json` can have the following format as well:
+    // {
+    //  "eos_token": {
+    //     ...
+    //     "content": "</s>",
+    //     ...
+    //   },
+    // }
+    // Once we have the eos token content, we can get its id from the tokenizer.
+    //
+    // For example:
+    // - https://huggingface.co/hf-internal-testing/llama-tokenizer/blob/main/tokenizer_config.json
+    EosTokenLocation {
+        file: "tokenizer_config.json",
+        location: EosTokenField::Object,
+    },
+];
+
+/// `Id` kind of `EosTokenField`, when `eos_token_id` provided as an id.
+#[derive(Debug, Serialize, Deserialize)]
+struct Id {
+    eos_token_id: u64,
+}
+
+/// `Value` kind of `EosTokenField`, when `eos_token` provided as a text, so that its id
+/// will be fetched from the tokenizer.
+#[derive(Debug, Serialize, Deserialize)]
+struct Value {
+    eos_token: String,
+}
+
+/// `Object` kind of `EosTokenField`, when `eos_token` provided as a `Content`.
+#[derive(Debug, Serialize, Deserialize)]
+struct Object {
+    eos_token: Content,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Content {
```

Comment:
The other structs in this module is documented, let's document this one as well, for consistency.